pkg/flowaggregator/certificate/provider.go

```diff
@@ -0,0 +1,586 @@
+		// Client cert could not be validated by the CA, regenerate client based on CA.
+		caCert, caKey, err := pemToCertKey(caCertPEM, caKeyPEM)
+		if err != nil {
+			return err
+		}
+
+		clientCertPEM, clientKeyPEM, err := GenerateCertKey(caCert, caKey, p.validFrom(), false, "")
+		if err != nil {
+			return fmt.Errorf("failed to generate client cert: %w", err)
+		}
+		if err := p.syncCertificateSecrets(ctx, clientSecretName, clientCertPEM, clientKeyPEM, clientSecret); err != nil {
+			if !errors.IsAlreadyExists(err) && !errors.IsConflict(err) {
+				return fmt.Errorf("failed to sync client Secret: %w", err)
+			}
+		}
+
+		klog.Info("Client certificates were rotated")
+	}
+	return nil
+}
+
+func (p *Provider) validFrom() time.Time {
+	// We set certificates as valid starting a hour earlier due to potential time skew on nodes.
+	return p.clock.Now().Add(-time.Hour)
+}
+
+func (p *Provider) shouldRotateCertificate(certPEM []byte, keyPEM []byte) bool {
+	if len(certPEM) == 0 || len(keyPEM) == 0 {
+		return true
+	}
+
+	cert, err := tls.X509KeyPair(certPEM, keyPEM)
+	if err != nil {
+		klog.ErrorS(err, "certificate not signed by the correct private key")
+		return true
+	}
+
+	remainingDuration := cert.Leaf.NotAfter.Sub(p.clock.Now())
+	if remainingDuration < minValidDuration {
+		klog.InfoS("The remaining duration of the TLS certificate and key is less than min valid duration", "remaining", remainingDuration, "min", minValidDuration)
+		return true
+	}
+	return false
+}
+
+func (p *Provider) getSecret(name string) (certPEM []byte, keyPEM []byte, secret *v1.Secret, err error) {
+	res, err := p.secretLister.Secrets(p.namespace).Get(name)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("error getting Secret %q: %w", caConfigMapName, err)
+	}
+
+	return res.Data[secretCertKey], res.Data[secretKeyKey], res, nil
+}
+
+func (p *Provider) syncCertificateSecrets(ctx context.Context, name string, cert, key []byte, secret *v1.Secret) error {
+	klog.InfoS("Syncing Secret", "name", name)
+
+	var res *v1.Secret
+	if secret == nil {
+		res = &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: p.namespace,
+			},
+			Type: v1.SecretTypeTLS,
+		}
+	} else {
+		res = secret.DeepCopy()
+	}
+
+	res.Data = map[string][]byte{
+		"tls.crt": cert,
+		"tls.key": key,
+	}
+	if secret != nil {
+		if _, err := p.k8sClient.CoreV1().Secrets(p.namespace).Update(ctx, res, metav1.UpdateOptions{}); err != nil {
+			return fmt.Errorf("failed to update Secret %s: %v", clientSecretName, err)
+		}
+	} else {
+		if _, err := p.k8sClient.CoreV1().Secrets(p.namespace).Create(ctx, res, metav1.CreateOptions{}); err != nil {
+			return fmt.Errorf("failed to create Secret %s: %v", clientSecretName, err)
+		}
+	}
+	return nil
+}
+
+func (p *Provider) syncCAConfigMap(ctx context.Context, name string, cert []byte) error {
+	var caConfigMap *v1.ConfigMap
+	res, err := p.configMapLister.ConfigMaps(p.namespace).Get(caConfigMapName)
```

Comment:
It's a bit confusing because your naming convention is switched compared to `syncCertificateSecrets` above. In `syncCertificateSecrets`, `res` is the resource to create / update, while here `res` in the current resource

Maybe you could use `currentSecret` / `desiredSecret` and `configMap` / `desiredConfigMap`, or something along those lines.